### PR TITLE
Allows ABAC authorizor to load policy file from specified etcd server

### PR DESCRIPTION
**Pull request by huang195**
Addresses #22988 

Allows ABAC authorizor to load a policy file directly from a specified etcd server. This PR is backward compatible.

Usage: specifying the following flag allows the policy file to be loaded from an etcd server at `http://10.143.100.209:4001` with a key `/security/abac-policy`

```
--authorization-policy-file=etcd@http://10.143.100.209:4001/security/abac-policy
```

### CODE REVIEW

**Comment by huang195** [comment]
@liggitt can you please take a look, thanks!

File: cmd/kube-apiserver/app/options/options.go [hunk 1]
```diff
@@ -218,7 +218,7 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&s.ServiceAccountLookup, "service-account-lookup", s.ServiceAccountLookup, "If true, validate ServiceAccount tokens exist in etcd as part of authentication.")
 	fs.StringVar(&s.KeystoneURL, "experimental-keystone-url", s.KeystoneURL, "If passed, activates the keystone authentication plugin")
 	fs.StringVar(&s.AuthorizationMode, "authorization-mode", s.AuthorizationMode, "Ordered list of plug-ins to do authorization on secure port. Comma-delimited list of: "+strings.Join(apiserver.AuthorizationModeChoices, ","))
-	fs.StringVar(&s.AuthorizationConfig.PolicyFile, "authorization-policy-file", s.AuthorizationConfig.PolicyFile, "File with authorization policy in csv format, used with --authorization-mode=ABAC, on the secure port.")
+	fs.StringVar(&s.AuthorizationConfig.PolicyFile, "authorization-policy-file", s.AuthorizationConfig.PolicyFile, "File with authorization policy in csv format, used with --authorization-mode=ABAC, on the secure port. Alternatively, file can be loaded from etcd by specifying etcd@http://<ip>:<port>/path/to/file. Multiple etcd servers can be specified separated using comma.")
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go]: I don't think we should overload one flag to either point to a file or an etcd connection string + path.
Could this not be accomplished with a custom bring-up script that uses etcdctl to pull the shared abac policy file from etcd to a local file prior to starting the api server?

File: pkg/auth/authorizer/abac/abac.go [hunk 2]
```diff
@@ -51,17 +56,78 @@ func (p policyLoadError) Error() string {
 
 type policyList []*api.Policy
 
+func getPolicyFile(policyFile string) (*bufio.Scanner, error) {
+
+	arr := strings.Split(policyFile, "@")
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: files can contain `@` symbols

**Comment by liggitt** [comment]
I'm not convinced loading from a second etcd instance is something that needs built-in support...

**Comment by huang195** [comment]
I can fix the problems pointed out above.

The problem we're having now is there is no authorization module that allows even very basic authorization capabilities, such as adding users, updating user policy, revoking user privilege, etc. without having to reboot the apiserver. (This is kind of a bad out-of-box UX, IMO) I think it's mentioned that ABAC is viewed as a way to only bootstrap. However, the more sophisticated way of managing authorization is still under discussion, and probably won't be available any time soon. This doesn't leaves current users any practical choice in the mean time, so we wanted to explore ABAC to see if it can be extended just a little bit to cover the above mentioned use cases without adding much code. It seems it is possible.

When ABAC policies are stored in etcd, changes to add/remove/update user privileges can be watched and reloaded by the ABAC authorization module (I haven't added the code to watch for changes, but this can be easily done). ABAC isn't sophisticated and it's kind of messy when dealing with user policies, but at least it's a stop-gap solution until something becomes available in the future.

**Comment by liggitt** [comment]
As a short-term solution, I'd rather see an option to watch/reload the ABAC file than tightly coupling to a second etcd server.

**Comment by huang195** [comment]
Let me take a look at the file watch/reload functions for ABAC. Any preference on https://github.com/fsnotify/fsnotify, which seems to support couple different OSes, or https://godoc.org/golang.org/x/exp/inotify, which is Linux-only.

**Comment by liggitt** [comment]
closing, recommendation for dynamic authz is to move to the rbac authorizer

**Pull request closed without merging**